### PR TITLE
Add Tensor.multinomial 

### PR DESCRIPTION
**Pull request by marcellofuschi**
Only with replacement (for now)

### CODE REVIEW

**Comment by geohot** [comment]
Locked the bounty for you

**Comment by marcellofuschi** [comment]
Fixed the first assert — the function now works with both 1D and 2D inputs. You cannot pass 3D or higher, just like torch

As for the second assert, sampling without replacement is trickier to implement lazily, so I'll need to think about it. Meanwhile, I'll set the PR as ready in case it's already useful just with replacement.

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -216,6 +216,19 @@ def kaiming_normal(*shape, a:float = 0.01, **kwargs) -> Tensor:
     std = math.sqrt(2.0 / (1 + a ** 2)) / math.sqrt(prod(shape[1:]))
     return Tensor.normal(*shape, mean=0.0, std=std, **kwargs)
 
+  @staticmethod
+  def multinomial(p: Tensor, num_samples: int, replacement: bool = False) -> Tensor:
+    assert p.ndim <= 2, "p must be 1 or 2 dim"
+    assert replacement, "supported only with replacement"
```
Comment (geohot) [inline tinygrad/tensor.py:222]: If num_samples is 1, is replacement=false okay?

Comment (marcellofuschi) [reply tinygrad/tensor.py:222]: yep, I'll allow it

**Comment by imaolo** [comment]
Does any runtime checking need to happen (looking at torch.multinomial with negative values). I don't think there are OPs that raise exceptions, atleast I don't know about them.

File: test/test_randomness.py [hunk 2]
```diff
@@ -96,6 +96,22 @@ def test_kaiming_normal(self):
...
+      if torch_res.ndim == 1:
+        tiny_res = tiny_res.unsqueeze(0)
+        torch_res = torch_res.unsqueeze(0)
+      for i in range(torch_res.shape[0]):
+        self.assertTrue(equal_distribution(lambda *_: tiny_res[i], lambda _: torch_res[i]))
+    _check_with_torch(p=[0.231, 0., 1., 0.5], num_samples=1000, replacement=True)
+    _check_with_torch(p=[[0.453, 0., 1., 0.81], [0.1, 0.8, 0., 0.1]], num_samples=1000, replacement=True)
+    # without-replacement isn't supported, unless taking only one sample
+    self.assertRaises(AssertionError, lambda: Tensor.multinomial(Tensor([0.5, 0.5]), num_samples=100, replacement=False))
+    _check_with_torch(p=[[0.5, 0.5], [0.99, 0.01]], num_samples=1, replacement=False)
```
Comment (geohot) [inline test/test_randomness.py:113]: Umm, doesn't this just fail 1% of the time?

Comment (marcellofuschi) [reply test/test_randomness.py]: It never fails actually, because for `num_samples=1` the ks test isn't reliable. That line was just checking that no error was raised when passing `num_samples=1, replacement=False`.

I've just changed the test code to make it more reliable by accumulating the results of repeated experiments.

Regarding the reliability of `equal_distribution` in general, I wrote a counterexample to test that the samples are considered from different distributions when they should. I ran it 1000 times and it always passed. I don't know if we should keep it in the code since the other tests in the file have no such thing.

Comment (geohot) [reply test/test_randomness.py]: Counterexample is great! Confirms the tests are actually testing something

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -216,6 +216,19 @@ def kaiming_normal(*shape, a:float = 0.01, **kwargs) -> Tensor:
     std = math.sqrt(2.0 / (1 + a ** 2)) / math.sqrt(prod(shape[1:]))
     return Tensor.normal(*shape, mean=0.0, std=std, **kwargs)
 
+  @staticmethod
```
Comment (geohot) [inline tinygrad/tensor.py:219]: This shouldn't be a staticmethod, and the first arg should be self.

Fixes `p.multinomial` and is a line shorter

**Comment by marcellofuschi** [comment]
Ok, I removed staticmethod and changed the tests accordingly.

Plus, I found a bug with an edge case (input 2D tensor with only 1 row should still return 2D). I added an assert and fixed it.

**Comment by geohot** [comment]
Sweet, can pay bounty via PayPal or USDC. E-mail george@tinygrad.org (or DM on discord) with your info to claim!
